ouroboros-consensus/src/Ouroboros/Consensus/MiniProtocol/ChainSync/Client.hs

```diff
@@ -709,113 +757,145 @@ chainSyncClient mkPipelineDecision0 tracer cfg
+          preCheck :: m ()
+          preCheck = do
+            -- Reject the block if invalid
+            isInvalidBlock <- atomically $ forgetFingerprint <$> getIsInvalidBlock
+            whenJust (isInvalidBlock hdrHash) $ \reason ->
+              disconnect $ InvalidBlock hdrPoint reason
+
+          -- | Check that the candidate intersects with our chain and that the
+          -- header is in forecast range from the intersection.
+          intersectCheck :: KnownIntersectionState blk -> m (IntersectCheck blk)
+          intersectCheck kis = do
+            -- Get the ledger view required to validate the header
+            -- NOTE: This will block if we are too far behind.
+            result <- atomically $ do
+            -- Before obtaining a 'LedgerView', we must find the most recent
+            -- intersection with the current chain. Note that this is cheap when
+            -- the chain and candidate haven't changed.
+              mKis' <- intersectsWithCurrentChain kis
+              case mKis' of
+                Nothing -> return $ Right NoLongerIntersects
+                Just kis'@KnownIntersectionState { mostRecentIntersection } -> do
+                  -- We're calling 'ledgerViewForecastAt' in the same STM transaction
+                  -- as 'intersectsWithCurrentChain'. This guarantees the former's
+                  -- precondition: the intersection is within the last @k@ blocks of
+                  -- the current chain.
+                  forecast <-
+                    maybe
+                    (error $
+                     "intersection not within last k blocks: " <> show mostRecentIntersection)
+                    (ledgerViewForecastAt (configLedger cfg) . ledgerState)
+                    <$> getPastLedger mostRecentIntersection
+
+                  case runExcept $ forecastFor forecast (blockSlot hdr) of
+                    -- The header is too far ahead of the intersection point with our
+                    -- current chain. We have to wait until our chain and the
+                    -- intersection have advanced far enough. This will wait on
+                    -- changes to the current chain via the call to
+                    -- 'intersectsWithCurrentChain' before it.
+                    Left OutsideForecastRange{} -> do
+                      -- If we reach a block which is outside of the forecast range,
+                      -- it means that it is outside of the genesis window as its
+                      -- length is the same as the forecast range.
+                      --
+                      -- This is a corner case that could be avoided if we switched to
+                      -- block-number based Ouroboros Genesis.
+                      t <- readTVar varCandidate
+                      if completeness t == FragmentComplete
+                        then
+                          -- as the value is already updated, we retry and let
```

Comment:
I wonder if it'd be useful to explain this further. What does it mean that the value is already updated in this context? What are we waiting for (what's the "something interesting" we're waiting for).